### PR TITLE
fix(policy): fix reapply policy after the dataset becomes valid again

### PR DESCRIPTION
**Pull request by mclcavalcante**
fix reapplying policy when the dataset becomes valid again after the sink which was linked is removed and a new one is added

### CODE REVIEW

**Review by lpegoraro — DISMISSED** [review]
LGTM

**Review by weyrick — CHANGES_REQUESTED** [review]
see thread https://ns1labs.slack.com/archives/C01UNG6UK37/p1653071246986939

**Review by lpegoraro — APPROVED** [review]
LGTM

**Review by weyrick — CHANGES_REQUESTED** [review]

File: policies/redis/producer/events.go [hunk 1]
```diff
@@ -53,6 +54,15 @@ type removeDatasetEvent struct {
...
 }
 
+type updateDatasetEvent struct {
+	id           string
+	ownerID      string
+	agentGroupID string
+	datasetID    string
+	policyID     string
+	timestamp    time.Time
+}
```
Comment (weyrick) [inline policies/redis/producer/events.go:64]: see my other comment: add `valid`, `turnedValid`, `turnedInvalid`

File: fleet/comms.go [hunk 2]
```diff
@@ -69,6 +71,68 @@ type fleetCommsService struct {
...
+	dataset, err := svc.policyClient.RetrieveDataset(ctx, &pb.DatasetByIDReq{
+		DatasetID: datasetID,
+		OwnerID:   ag.MFOwnerID,
+	})
+	if err != nil {
+		return err
+	}
+
+	var action string
+	if dataset.Valid {
```
Comment (weyrick) [inline fleet/comms.go:94]: change to TurnedValid

File: fleet/comms.go [hunk 3]
```diff
@@ -69,6 +71,68 @@ type fleetCommsService struct {
...
+	if err != nil {
+		return err
+	}
+
+	var pdata interface{}
+	if err := json.Unmarshal(p.Data, &pdata); err != nil {
+		return err
+	}
+
+	dataset, err := svc.policyClient.RetrieveDataset(ctx, &pb.DatasetByIDReq{
```
Comment (weyrick) [inline fleet/comms.go:85]: we shouldn't need to call this, it should get all information it needs from the redis stream

**Review by weyrick — CHANGES_REQUESTED** [review]

File: fleet/redis/consumer/streams.go [hunk 4]
```diff
@@ -147,6 +152,32 @@ func (es eventStore) handleDatasetRemove(ctx context.Context, e removeDatasetEve
...
+		datasetID:     read(event, "dataset_id", ""),
+		policyID:      read(event, "policy_id", ""),
+		valid:         readBool(event, "valid", false),
+		turnedValid:   readBool(event, "turned_valid", false),
+		turnedInvalid: readBool(event, "turned_invalid", false),
+	}
+}
+
+func (es eventStore) handleDatasetUpdate(ctx context.Context, e updateDatasetEvent) error {
+	ag, err := es.fleetService.ViewAgentGroupByIDInternal(ctx, e.agentGroupID, e.ownerID)
```
Comment (weyrick) [inline fleet/redis/consumer/streams.go:169]: this should be done in the block below so that we avoid a database lookup unless we really need it

**Review by weyrick — CHANGES_REQUESTED** [review]
almost there, one minor optimization and i think we are ready

**Review by weyrick — APPROVED** [review]

**Pull request merged**